pytensor/tensor/subtensor.py

```diff
@@ -145,15 +151,34 @@ def as_index_constant(
         )
     elif isinstance(a, int | np.integer):
         return ps.ScalarConstant(ps.int64, a)
-    elif not isinstance(a, Variable):
-        return as_tensor_variable(a)
-    else:
+    elif isinstance(a, Variable):
         return a
+    return as_tensor_variable(a)
+
+
+@overload
+def as_index_literal(idx: int | np.integer) -> int | np.integer: ...
+
+
+@overload
+def as_index_literal(idx: None) -> None: ...
+
+
+@overload
+def as_index_literal(idx: slice) -> slice: ...
+
+
+@overload
+def as_index_literal(idx: Constant) -> int | np.integer | None: ...
+
+
+@overload
+def as_index_literal(idx: Variable) -> int | np.integer | slice | None: ...
 
 
 def as_index_literal(
-    idx: Variable | slice | None,
-) -> int | slice | None:
+    idx: int | np.integer | Variable | slice | None,
```

Comment:
Maybe be explicit about the valid Variable types?
```suggestion
    idx: int | np.integer | ScalarVariable | TensorVariable | slice 
    | SliceVariable | None | NoneConst, 
```

Reply:
no longer applies